lib/Runtime/PlatformAgnostic/Intl.h

```diff
@@ -84,34 +84,74 @@ namespace Intl
         // Note: Cleanup() is called by, e.g., AutoIcuJsObject
     };
 
+    template<typename TResource>
+    class IcuCObject : public IPlatformAgnosticResource
+    {
+    public:
+        typedef void (__cdecl *CleanupFunc)(TResource *);
+
+    private:
+        TResource *resource;
+        CleanupFunc cleanupFunc;
+
+    public:
+        IcuCObject(TResource *resource, CleanupFunc cleanupFunc) :
+            resource(resource),
+            cleanupFunc(cleanupFunc)
+        {
+
+        }
+
+        TResource *GetInstance()
+        {
+            return resource;
+        }
+
+        void Cleanup() override
+        {
+            cleanupFunc(resource);
+            resource = nullptr;
```

Comment:
If the contract of `IcuCObject` is that after `Cleanup()` it should never be accessed again (and specifically that `Cleanup()` should be invoked as part of a GC finalize meaning there are no lingering references) then you could `delete this;` to avoid the memory leak you mention later.